```{r load_libraries}
library(data.table)
library(tidyverse)
library(rairtable)
library(tinytable)
library(ggh4x)
library(cluster)
library(gridExtra)
```

```{r load_vcf}

# List all vcf.gz files in the folder
vcf_files <- list.files("data/vcf", pattern = "\\.vcf\\.gz$", full.names = TRUE)

# Function to process a single VCF file
process_vcf <- function(file_path) {
  fread(file_path, sep = "\t") %>%
    as_tibble() %>%
    rename(start = 1) %>%
    filter(!str_starts(start, "#")) %>%
    separate(
      start,
      into = c("contig", "position", "id", "ref", "alt", "qual", "filt", "info", "format", "genotype"),
      sep = "\t"
    ) %>%
    mutate(variant = str_c(contig, "_", position),
           genome = str_extract(contig, "^[^@]+"),
           DP = as.numeric(str_extract(info, "(?<=DP=)[^;]+")),
           MQ = as.numeric(str_extract(info, "(?<=MQ=)[^;]+")),
           qual=as.numeric(qual)) %>% 
    filter(as.numeric(qual) > 20) %>%
    arrange(variant)
}

# Process all VCF files and store them in a list of tibbles
vcf_list <- lapply(vcf_files, process_vcf)

# Name the list elements by file name (without full path)
names(vcf_list) <- basename(vcf_files) %>% str_remove(".vcf.gz")

```

```{r load_metadata}
microsample_metadata <- airtable("4-MSE-Info", "appKakM1bnKSekwuW") %>% 
  read_airtable(., fields = c("ID","cryosection_text","Ycoord","Xcoord","SampleType"), id_to_col = TRUE) %>% 
  rename(microsample=ID,cryosection=cryosection_text) %>% 
  filter(microsample %in% names(vcf_list)) %>% 
  select(-airtable_record_id) %>% 
  unnest(cols = c(Xcoord, Ycoord)) %>% 
  mutate(animal = str_sub(cryosection, 1, 4))

microsample_pairwise <- microsample_metadata %>%
  filter(!is.na(cryosection)) %>% 
  group_by(cryosection) %>%  
  summarise(pairwise_distances = list({
    coords <- pick(microsample, Xcoord, Ycoord)
    coord_matrix <- coords %>%
      column_to_rownames("microsample") %>%          # Use microsample as row names
      select(Xcoord, Ycoord) %>%
      as.matrix()
    dist(coord_matrix) %>%                           # Calculate pairwise distances
      as.matrix() %>%                                # Convert to matrix
      as_tibble(rownames = "microsample_1") %>%      # Retain sample names as row names
      pivot_longer(
        cols = -microsample_1,
        names_to = "microsample_2",
        values_to = "distance"
      )
  }), .groups = "drop")               
```

#### Taxonomy
This is the raw taxonomy table generated by GTDBtk, which is simplified for downstream analyses.
```{r load_taxonomy, message=FALSE, eval=FALSE}
genome_taxonomy <- read_tsv("data/genome_taxonomy.tsv") %>%
  rename(genome = user_genome) %>%
  mutate(genome = str_replace_all(genome,"\\.fa", "")) %>%
  separate(classification, c("domain","phylum","class","order","family","genus","species"),  sep =";") %>%
  select(genome,domain,phylum,class,order,family,genus,species)
```

#### Genome quality
Quality properties of the genomes. 
```{r load_quality, message=FALSE, eval=FALSE}
genome_quality <- read_tsv("data/genome_quality.tsv") %>%
  rename(genome = 1) %>%
  mutate(genome = str_replace_all(genome,"\\.fa", "")) %>%
  select(genome, Completeness, Contamination, Coding_Density, Genome_Size) %>%
  rename(completeness=Completeness,contamination=Contamination,coding_density=Coding_Density,length=Genome_Size)
```

#### Merged metadata object
Merge taxonomy, length and quality information
```{r create_genomemetadata, message=FALSE, eval=FALSE}
genome_metadata <- genome_taxonomy %>%
  left_join(genome_quality,by=join_by(genome==genome)) #join quality
```

## Count data

```{r load_count_data, message=FALSE, eval=FALSE}
read_counts <- read_tsv("data/coverm_genome_REF0009-mgg-pbdrep.count.tsv") %>%
  rename(genome = 1) %>%
  pivot_longer(!genome, names_to = "data", values_to="counts") %>%
  mutate(sample = substr(data, 1, 7)) %>%
  group_by(genome,sample) %>%
  summarise(counts=sum(counts), .groups="drop") %>%
  pivot_wider(names_from="sample", values_from="counts")

readlength=150 #change if sequencing read length is different
genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r load_annotations}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>% 
  rename(gene=1,genome=2,contig=3)
```

```{r filtering}
vcf_list_filt <- map(vcf_list, ~ .x %>%
  filter(qual > 50,
         DP > 10,
         MQ > 30)) %>% 
  keep(~ nrow(.x) > 0) #remove empty genomes

#List of genomes
genomes <- vcf_list_filt %>%
  purrr::map(~ .x$genome) %>% 
  unlist() %>%
  unique()    
```


```{r stats}
depth_stats <- map_dfr(vcf_list_filt, ~ .x %>%
  group_by(genome) %>%
  summarise(depth = mean(DP)), .id = "sample") %>%
  pivot_wider(
    names_from = sample,
    values_from = depth,
    values_fill = 0
  )

snp_stats <- map_dfr(vcf_list_filt, ~ .x %>%
  group_by(genome) %>%
  summarise(n_variants = n()), .id = "sample") %>%
  pivot_wider(
    names_from = sample,
    values_from = n_variants,
    values_fill = 0
  )

```

Select abundant and widespread genomes for strain analyses

```{r genome_selection}
genome_selection_table <- depth_stats %>% 
  pivot_longer(!genome, names_to = "sample", values_to = "value") %>% 
  group_by(genome) %>% 
  summarise(mean=mean(value),
           sd=sd(value)) %>% 
  arrange(-mean,sd)

genome_selection <- genome_selection_table$genome[1:5]

depth_stats %>% 
  filter(genome %in% genome_selection) %>% 
  tt()

#GPB:bin_000088 - Caccomorpha excrementavium
```

```{r binary_analysis}

clusters_vector <- c()
silhouettes_vector <- c()

for(i in c(75:length(genomes))){
cat(str_c(i,"\n"))
vcf_list_genome <- purrr::map(vcf_list_filt, ~ .x %>%
  filter(genome == genomes[i])) %>% 
  keep(~ nrow(.x) > 0)

genome_annotations_filter <- genome_annotations %>% 
  filter(genome == genomes[i]) %>% 
  select(contig,start_position,end_position,gene,ko_id)

variants <- map_dfr( vcf_list_genome, 
  ~ .x %>% select(contig, position)) %>% 
  unique() %>% 
  mutate(position=as.numeric(position))

variant_annotations <- variants %>%
  left_join(
    genome_annotations_filter, 
    by = "contig",
    relationship = "many-to-many"
  ) %>%
  filter(
    position >= start_position & position <= end_position
  ) %>%
  unique() %>% 
  mutate(variant=str_c(contig,"_",position)) %>% 
  select(variant, gene, start_position, end_position, ko_id)
  
binary_genome <- map_dfr(vcf_list_genome, ~ .x %>%
  select(variant), .id = "sample") %>%  
  distinct() %>% 
  mutate(present = 1) %>% 
  pivot_wider(
    names_from = sample,
    values_from = present, 
    values_fill = list(present = 0))

binary_matrix <- binary_genome %>%
  column_to_rownames("variant") %>%
  as.matrix()

maxstrain <- if(ncol(binary_matrix)<11){maxstrain=ncol(binary_matrix)-2}else{maxstrain=10}

if(ncol(binary_matrix)>2 & nrow(binary_matrix)>1 ){
    binary_silhouette <- purrr::map(1:maxstrain, function(k) {
      pam(binary_matrix, k = k, metric = "euclidean")$silinfo$avg.width
      }) %>% 
      purrr::map(., ~ if (is.null(.x)) 0 else .x) %>% 
      unlist()
    
    binary_strains <- purrr::map(which.max(binary_silhouette), function(k) {
      pam(binary_matrix, k = k, metric = "euclidean")$clustering
      }) %>% 
      purrr::map(., ~ if (is.null(.x)) 0 else .x) %>% 
      unlist()
    
    clusters_vector <- c(clusters_vector, which.max(binary_silhouette))
    silhouettes_vector <- c(silhouettes_vector,max(binary_silhouette))
    
    clustering_binary_genome <- binary_genome %>%
        column_to_rownames("variant") %>%
        as.matrix() %>%
        t() %>% 
        dist(method = "euclidean") %>% 
        hclust(method = "average") 
          
    
    #Plot figures
    pdf(str_c("figures/hclust_",gsub(":","_",genomes[i]),".pdf"),width=10,height=6)
      plot(clustering_binary_genome, main=genomes[i])
    dev.off()
    
    represented_samples <- genome_counts %>%
      pivot_longer(!genome,names_to = "microsample", values_to = "coverage") %>% 
      filter(genome==genomes[i],
             microsample %in% names(binary_genome)[-1]) %>% 
      filter(coverage>1)
    
    variant_figure <- binary_genome %>% 
      pivot_longer(!variant,names_to = "microsample",values_to = "genotype") %>% 
      mutate(sorting = as.numeric(str_extract(variant, "[^_]+$"))) %>% 
      arrange(sorting) %>% 
      mutate(position = factor(sorting)) %>% 
      mutate(sample = factor(microsample,levels=clustering_binary_genome$labels[clustering_binary_genome$order])) %>% 
      #left_join(represented_samples,by="microsample") %>% 
      mutate(genotype = factor(genotype)) %>% 
      left_join(microsample_metadata,by="microsample") %>% 
      filter(!is.na(cryosection)) %>% 
      left_join(variant_annotations,by="variant") %>% 
      ggplot(aes(x=fct_reorder(position, sorting),y=microsample,fill=genotype)) +
        geom_tile() +
        scale_fill_manual(values=c("#ffffff","#99cc00")) + 
        facet_nested(animal+cryosection~gene,scales="free",space="free") + 
        theme(axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              strip.background.y = element_blank(),
              strip.text.y = element_blank(),
              legend.position="none")
    
    coverage_figure <- genome_counts %>%
      pivot_longer(!genome,names_to = "microsample", values_to = "coverage") %>% 
      filter(genome==genomes[i],
             microsample %in% names(binary_genome)[-1]) %>% 
      left_join(microsample_metadata,by="microsample") %>% 
      filter(!is.na(cryosection)) %>% 
      mutate(abundance="Abundance") %>% 
      ggplot(aes(x=coverage,y=microsample)) +
        geom_col() +
        facet_nested(animal+cryosection~abundance,scales="free",space="free") +
        theme(axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank())
    
    pdf(str_c("figures/map_",gsub(":","_",genomes[i]),".pdf"),width=10,height=6)
    grid.arrange(
        variant_figure,
        coverage_figure,
        ncol = 2, 
        widths=c(5,1)
      )
     dev.off()
    
  }
}

strain_table <- tibble(genome=genomes,strains=clusters_vector,silhouette=silhouettes_vector)

```

```{r binary_plot}
clustering_binary_genome <- binary_genome %>%
  column_to_rownames("variant") %>%
  as.matrix() %>%
  t() %>% 
  dist(method = "euclidean") %>% 
  hclust(method = "average") 

plot(clustering_binary_genome)
```

```{r depth_analysis}
depth_genome <- map_dfr(vcf_list_genome, ~ .x %>%
  select(variant, DP), .id = "sample") %>%
  distinct() %>%
  pivot_wider(
    names_from = sample,
    values_from = DP,
    values_fill = list(DP = 0)
  )

depth_matrix <- depth_genome %>%
  column_to_rownames("variant") %>%
  as.matrix()

depth_silhouette <- purrr::map(1:10, function(k) {
  pam(depth_matrix, k = k, metric = "euclidean")$silinfo$avg.width
  }) %>% 
  purrr::map(., ~ if (is.null(.x)) 0 else .x) %>% 
  unlist()

depth_strains <- purrr::map(which.max(depth_silhouette), function(k) {
  pam(binary_matrix, k = k, metric = "euclidean")$clustering
  }) %>% 
  purrr::map(., ~ if (is.null(.x)) 0 else .x) %>% 
  unlist()
```

```{r depth_plot}
clustering_depth_genome <- depth_genome %>%
  column_to_rownames("variant") %>%
  as.matrix() %>%
  t() %>% 
  dist(method = "euclidean") %>% 
  hclust(method = "average") 

plot(clustering_depth_genome)
```


## Heatmap

```{r binary_heatmap}
binary_genome %>% 
  pivot_longer(!variant,names_to = "microsample",values_to = "genotype") %>% 
  mutate(sorting = as.numeric(str_extract(variant, "[^_]+$"))) %>% 
  arrange(sorting) %>% 
  mutate(position = factor(sorting)) %>% 
  mutate(sample = factor(microsample, levels=clustering_binary_genome$labels[clustering_binary_genome$order])) %>% 
  mutate(genotype = factor(genotype)) %>% 
  left_join(microsample_metadata,by="microsample") %>% 
  ggplot(aes(x=fct_reorder(position, sorting),y=sample,fill=genotype)) +
    geom_tile() +
    scale_fill_manual(values=c("#ffffff","#99cc00")) + 
    facet_nested(animal+cryosection~.,scales="free",space="free") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r depth_heatmap}
depth_genome %>%
  mutate(across(
    .cols = -variant,                 # Exclude the `variant` column
    .fns = ~ . / mean(., na.rm = TRUE) # Scale by column mean
  )) %>% 
  pivot_longer(!variant,names_to = "microsample",values_to = "depth") %>% 
  mutate(sorting = as.numeric(str_extract(variant, "[^_]+$"))) %>% 
  arrange(sorting) %>% 
  mutate(position = factor(sorting)) %>% 
  mutate(sample = factor(microsample, levels=clustering_binary_genome$labels[clustering_binary_genome$order])) %>% 
  mutate(depth = as.numeric(depth)) %>% 
  left_join(microsample_metadata,by="microsample") %>% 
  ggplot(aes(x=fct_reorder(position, sorting),y=sample,fill=depth)) +
    geom_tile() +
    facet_nested(animal+cryosection~.,scales="free",space="free") + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
